File: storage/db-core/src/main/java/com/coffee/gu/stamp/StampHistoryRepositoryImpl.java
package com.coffee.gu.stamp;

import com.querydsl.jpa.impl.JPAQueryFactory;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.List;

import static com.coffee.gu.stamp.QStampHistoryEntity.stampHistoryEntity;

@Component
public class StampHistoryRepositoryImpl implements StampHistoryRepository {

    private final JPAQueryFactory queryFactory;
    private final StampHistoryJpaRepository stampHistoryJpaRepository;

    public StampHistoryRepositoryImpl(JPAQueryFactory queryFactory, StampHistoryJpaRepository stampHistoryJpaRepository) {
        this.queryFactory = queryFactory;
        this.stampHistoryJpaRepository = stampHistoryJpaRepository;
    }

    @Override
    public StampHistory save(StampHistory stampHistory) {
        return stampHistoryJpaRepository.save(StampHistoryEntity.from(stampHistory)).toModel();
    }

    @Override
    public List<StampHistory> getWithinNMonths(String principalKey, LocalDateTime nMonthsAgo) {
        return queryFactory
                .selectFrom(stampHistoryEntity)
                .where(stampHistoryEntity.principalKey.eq(principalKey)
                        .and(stampHistoryEntity.createdAt.after(nMonthsAgo)))
                .orderBy(stampHistoryEntity.createdAt.desc())
                .fetch()
                .stream()
                .map(StampHistoryEntity::toModel)
                .toList();
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/stamp/StampHistoryRepositoryImpl.kt
package com.coffee.gu.stamp

import com.coffee.gu.stamp.QStampHistoryEntity.stampHistoryEntity
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository
import java.time.LocalDateTime

@Repository
class StampHistoryRepositoryImpl(
    private val queryFactory: JPAQueryFactory,
    private val stampHistoryJpaRepository: StampHistoryJpaRepository,
) : StampHistoryRepository {

    override fun save(stampHistory: StampHistory): StampHistory {
        return stampHistoryJpaRepository.save(StampHistoryEntity.from(stampHistory)).toModel()
    }

    override fun getWithinNMonths(principalKey: String, nMonthsAgo: LocalDateTime): List<StampHistory> {
        return queryFactory
            .selectFrom(stampHistoryEntity)
            .where(
                stampHistoryEntity.principalKey.eq(principalKey),
                stampHistoryEntity.createdAt.after(nMonthsAgo),
            )
            .orderBy(stampHistoryEntity.createdAt.desc())
            .fetch()
            .map { it.toModel() }
    }
}
